### ISSUE TITLE
Windows MaxPath Restriction

### ISSUE THREAD

**Issue by aersamkull** [composer/composer#4628]
Composer seems to be subject to the Windows MaxPath restriction. Windows does support paths longer than 260 paths (via special syntax). Other module loaders such as npm do support this.

**Comment by alcohol** [comment]
We are always open to pull requests. I personally have no experience developing on Windows so I wouldn't even know where to begin looking for details on this, other than Google.

**Comment by aersamkull** [comment]
Well, I do not know the source yet. Actually,  it is simple to fix. Whenever you write a file, 
do something like this:
```PHP
if(strlen($path)>=256 && DIRECTORY_SEPARATOR == '\\') {
    $path =  "\\?\" . $path;
}
```

**Comment by aersamkull** [comment]
Ou, relative paths longer than 256 had to be converted to absolute paths. But that is very unlikely

**Comment by alcohol** [comment]
Alright, well those are good pointers for anyone willing to work on this. I might even pick it up myself if I can find the time.

**Comment by alcohol** [comment]
One question though, since what Windows version is this syntax supported?

**Comment by alcohol** [comment]
I was able to find this:

> Maximum Path Length Limitation
>
In the Windows API (with some exceptions discussed in the following paragraphs), the maximum length for a path is MAX_PATH, which is defined as 260 characters. A local path is structured in the following order: drive letter, colon, backslash, name components separated by backslashes, and a terminating null character. For example, the maximum path on drive D is "D:\some 256-character path string<NUL>" where "<NUL>" represents the invisible terminating null character for the current system codepage. (The characters < > are used here for visual clarity and cannot be part of a valid path string.)
>> Note  File I/O functions in the Windows API convert "/" to "\" as part of converting the name to an NT-style name, except when using the "\\?\" prefix as detailed in the following sections.

More details can be found [here](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx).

**Comment by Seldaek** [comment]
As far as I know there is no support for long paths on windows in PHP.. Even php7 doesn't seem to fix it yet unfortunately. It sucks but AFAIK we can't do anything about it from our end. If you find a way though then it'd be great I agree. In the meantime I'll close this because I believe it should just be fixed in PHP really.

**Comment by aersamkull** [comment]
The bug report for PHP can be found here: https://bugs.php.net/bug.php?id=70943 .

**Comment by XedinUnknown** [comment]
FYI, bug reported to be fixed in PHP 7.1, albeit without mentioning the PR/commit which resolves the issue.

**Comment by Seldaek** [comment]
I can confirm that 7.1 fixes that test script in the bug report for me on windows, which is great news :)
